Narrow focus of proposal to module loading (#46)
tc39/proposal-compartments: README.md

* import mappers ([import-map][import-map]) like bundlers need to be able to
  collect transitive dependencies according to ECMAScript language and specific
  host behaviors.
  A ECMAScript native module loader interface would expedite evolution of import map
  runtimes in JavaScript.
* hot module replacement (HMR) systems (WebPack, SnowPack, &c), which need the
  ability to instantiate new module graphs when dependencies change and the
  ability to bequeath subgraphs to new graphs.
  * Node.js [defers][node-hmr] to ECMAScript to provide a module loader
    interface to aid HMR.
* persistent testing apparatuses ([Jest][jest]), because a persistent service
  reinstantiates whole module graphs to reconstruct tests and test subjects.
  * Jest currently resorts to exploiting Node.js's [vm][vm-context] module to
    instantiate separate realms and attempts ([and
    fails][jest-ses-interaction]) to provide the illusion of a single realm by
    patching client realms with some of the intrinsics of the host realm.
* emulators ([JSDom][jsdom]) in which the emulated artifact may need a separate
  module memo from the surrounding realm.

> * sub-realm sandboxes ([SES][ses] and [LavaMoat][lava-moat]) that virtualize

Reviewer: ```suggestion
* sub-realm sandboxes ([SES][ses-shim] and [LavaMoat][lava-moat]) that virtualize
```

    // if the `record` has a `true` `needsImportMeta` property.
    importMeta?: Object,
  }

  // Use a static module record from the compartment in which this compartment
  // was constructed.
  // If the compartment _loads_ the corresponding module, the static
  // module record will be synchronously available only if it is already
  // in the parent compartment's synchronous memo.
  // Otherwise, loading induces the parent compartment to load the module,
  // but does not induce the parent compartment to load the module's transitive
  // dependencies.
  // This compartment then resolves the shallow dependencies according
  // to its own resolveHook and loads the consequent transitive dependencies.
  // If the compartment _imports_ the module, the behavior is equivalent
  // to loading for the retrieved static module record or third-party static
  // module record.
  | {
    record: string,

    // Properties to copy to the `import.meta` of the resulting module instance.
    importMeta?: Object,
  }

  // FOR SHARED MODULE INSTANCES:

  // To create an alias to an existing module instance in this compartment:
  | {
    // A full specifier for another module in this compartment.

>     instance: string,

Reviewer 2: I'm wondering what's the difference between `instance: string` and `record: string`

    namespace: ^ModuleExportsNamespace,
  };

type CompartmentConstructorOptions = {
  // Every Compartment has a reference to a global environment record that in
  // turn contains a new globalThis object, global contour, and three
  // specialized intrinsic evaluators: eval, Function, and Compartment instances.
  // The new globalThis object contains a subset of the JavaScript language
  // intrinsics (to be defined in this proposal) and other globals must be
  // "endowed" to the compartment explicitly with the `globals` option.
  // All of these evaluators close over the compartment's global environment
  // record such that they evaluate code in that global environment.
  // When borrowGlobals is false, a the new Compartment gets a new global
  // environment record.
  // When borrowGlobals is true, the new Compartment will have the
  // same global environment record as associated with the Compartment
  // constructor used to construct the compartment.
  // To borrow the globals of an arbitrary compartment, use that compartment's
  // Compartment constructor, like
  // new compartment.globalThis.Compartment({ borrowGlobals: true }).

>   borrowGlobals: boolean,

Reviewer 2: How is it going to borrow globals meanwhile have the different evaluators?

Does it mean in this case, it is an exotic object?

Reviewer 3: Yeah I think this paragraph could use a re-wording. I believe in the global re-use, there is no different evaluators, they are the same.

Author: Right, global environment records own evaluators, compartments refer to a global environment record and the only difference is the guest’s is the same as the host’s or if the guest gets a new one.

I’ll seek to improve upon this working in a subsequent PR. That probably will involve moving the description out of the sketch.

// An `export name as alias` binding will have a property with the lexically bound name,
// whereas the module exports namespace will have a property with the alias.
interface ModuleEnvironmentRecord {}

// Bindings reflect the `import` and `export` statements of a module.
type Binding =
  { import: '*' | string, as?: string, from: string } |
  { export: '*' | string, as?: string, from?: string };

// Compartments support ECMAScript modules and linkage to other kinds of modules,
// notably allowing for JSON or WASM.
// These must provide an initializer function and may have bindings.
// The bindings correspond to the equivalent `import` and `export` declarations
// of an ECMAScript module.
type ThirdPartyStaticModuleRecord = {
  bindings?: Array<Binding>,
  // Initializes the module if it is imported.
  // Initialize may return a promise, indicating that the module uses
  // the equivalent of top-level-await.
  initialize(environment: ModuleEnvironmentRecord, meta: Object),
  // TODO decide upon a positive or negative form of `usesMeta` or `noMeta`.
  // Avoid calling `importMetaHook` for this module since it won't use the
  // `meta` property. 

>   noMeta?: boolean,

Author: We (@erights, @patrick-soquet, @phoddie, @mhofman, myself) converged on `needsImportMeta` in the positive, so by default, third-party static module record initializers would _not_ receive an `importMeta` argument (`undefined`).

>   // TODO consider whether to reflect top-level-await statically.

Author: @patrick-soquet verified that no special marker for async initializers is necessary. The emergent behavior of an async function with await should be identical to the equivalent ESM with a top-level await.

};

// Static module records are an opaque token representing the compilation
// of a module that can be reused across multiple compartments.
interface StaticModuleRecord {
  // Static module records can be constructed from source.
  // XS allows third-party module records and source descriptors to
  // be precompiled as well.
  constructor(source: string | { source: string } | ThirdPartyStaticModuleRecord);

  // Static module records reflect their bindings for information only.
  // Compartments use internal slots for the compiled code and bindings.
  bindings: Array<Binding>;
}

// A ModuleDescriptor captures a static module record and per-compartment metadata.
type ModuleDescriptor = 

>   // Using a string for a module descriptor indicates a static module record
>   // should be inherited from the parent compartment, but not the parent
>   // compartment's instance.
>   string

Author: @mhofman suggests that we should make module descriptors homogeneously objects.
```suggestion
  // Indicates a static module record
  // should be inherited from the parent compartment, but not the parent
  // compartment's instance.
  {
    inherit: string,
  }
```

Author: We (@erights, @patrick-soquet, @phoddie, @mhofman, myself) did not come to a conclusion either for “whether” or “with what name”.

>   string

Reviewer 2: If the incubator compartment does not have this module loaded, does it invoke the incubator's import hooks?

Author: I’ve added prose to answer your question. Please take a look at the descriptions for each kind of module descriptor.

  // A static module record descriptor produces a unique instance
  // of a module in a compartment.
  | {
    record: StaticModuleRecord | ThirdPartyStaticModuleRecord,

    // An optional alias.
    // For example, in a Node.js package compartment, '.' may imply a memo entry
    // for './index.js'.
    // TODO this design does not yet account for the possibility of multiple
    // aliases, which have not been necessary for a faithful imitation of Node.js.
    // That might be better accounted for with {compartment?, specifier}
    // module descriptors, below.

>     specifier?: string,

Author: We (@erights, @patrick-soquet, @phoddie, @mhofman, myself) agreed that this is a good direction and that we should remove `compartment.module`. We did not settle on the names of properties, but these will do tentatively.

    meta?: Object,
  }
  // As a shorthand for constructing a StaticModuleRecord,
  // a module descriptor can provide source.
  | {
    source: string,
    specifier?: string,
    meta?: Object,
  }
  // To refer to a module instance in another compartment, we need either to
  // have a module descriptor that can refer to it, or have a `module` method and
  // use module exports namespaces as opaque tokens to refer to modules in foreign
  // compartments.
  | {
    // A compartment instance.
    // We do not preclude the possibility that compartment is the same compartment,
    // as that is possible to achieve in moduleMapHook and loadHook, albeit not
    // in moduleMap.
    // TODO Idea: it may be sufficient to make this compartment instance optional,
    // defaulting to the current instance, in order to implement arbitrary
    // numbers of aliases to a single module instance, from one or many other
    // compartments, which would obviate the need for `Compartmnt.prototype.module`.
    compartment: Compartment,
    // The full specifier of the corresponding module instance in the other compartment.
    specifier: string,
  }
  | {
    namespace: ModuleExportsNamespace,

>   };

Author: We (@erights, @patrick-soquet, @phoddie, @mhofman, myself) haven’t settled on whether to continue allowing a bare module exports namespace (distinguished by brand check) as a kind of module descriptor. I’m still leaning toward not supporting the abbreviation.

We also did settle on allowing the `{ namespace }` module descriptor shape. The behavior for brand-check-passing module namespace objects would be special and allow lazy binding. The behavior for any other object would be to capture a snapshot of the object with a get own property names check and a get for every property, such that it realizes a module namespace object that exports a binding for all those properties with those constant values.

>   };

Reviewer 4: When I proposed the descriptor, I meant it to be instead of allowing the namespace directly. One of the goals was to preserve practical namespace transparency, without forcing existing membrane code to make a weird special case for module namespaces.

So do we agree that the descriptor is *instead* rather than *in addition to*?

MarkM: y

>   };

Reviewer 4: Where you say "lazy binding" above, did you mean "live binding"?

>   };

Author: Yes, I mean live binding.

I would like to support:

```js
new Compartment({
  moduleMap: {
    x: {
      namespace: {}
    }
  }
})
```

And also support:

```js
import * as x from 'x';
new Compartment({
  moduleMap: {
    x: {
      namespace: x
    }
  }
})
```

And I would prefer not to support (even though I agree it is convenient):

```js
import * as x from 'x';
new Compartment({
  moduleMap: { x }
})
```

>   };

Reviewer 4: Yes. I meant only to prohibit the last.

type CompartmentConstructorOptions = {
  // inherit, when true, indicates that this compartment shares the globals, global
  // contour, and evaluators of the surrounding compartment.
  // This would interact with the SES `lockdown` function: after lockdown,
  // a true would be incoherent and throw an exception.
  // When false, the compartment's globalThis would be populated only
  // with a subset of the language's intrinsics and the compartment's
  // specialized evaluators.
  // For this proposal to be fully specified, we would need to realize
  // the concept of shared intrinsics in anticipation of Lockdown.
  // In order for the module-loader Compartment to be useful for a
  // Lockdown-Compartment shim, we must disallow vendors from including
  // any shared intrinsics beyond those specified by the module loader proposal.
  inherit: boolean,

  // Globals to copy onto this compartment's unique globalThis.
  // Constructor options with globals and inherit: true would be incoherent and
  // effect an exception.

>   globals: Object,

Author: Nobody (between @erights and me) likes the name `inherit` for anything, so this is an open design question.

  loadNowHook?: (fullSpec: FullSpecifier) => ModuleDescriptor?;

  // A ModuleDescriptor can have a `meta` property.
  // The compartment assigns these properties over the true `import.meta`
  // object for the module instance.
  // That object in turn has a null prototype.
  // However some properties of `meta` are too expensive for a host
  // to instantiate for every module, like Node.js's `import.meta.resolve`,
  // which can be avoided for modules that never utter `import.meta`
  // in their sources.
  // This hook gets called for any module that utters `import.meta`,
  // so some work can be deferred until just before the compartment
  // initializes the module.
  importMetaHook?: (meta: Object) => void,
};

interface Compartment {
  // Note: This differs from the implementations of SES shim and Moddable's XS,
  // which accept two arguments before the options bag.
  // TODO Settle #5 before commiting to a single-options bag constructor.
  // We can effect a migration by initially requiring a symbol on the first
  // argument options bag, like `[Compartment.options]: true`, in order
  // afford new code that is forward compatible while still supporting

>   // legacy argument patterns.

Author: We (@erights, @patrick-soquet, @phoddie, @mhofman, myself) agreed to collapse `Compartment` to a single options bag argument!

  // persists to the next.
  // Subsequent proposals might afford other modes, like sloppy globals mode.
  // TODO is sloppyGlobalsMode only sensible in the context of the shim?
  evaluate(source: string): any;

  // load causes a compartment to load module descriptors for the
  // transitive dependencies of a specified module into its
  // memo, but does not initialize any modules.
  // The load function is useful for tools like bundlers and importmap
  // generators that load a module graph in an emulated host environment
  // but cannot and should not emulate evaluation.
  async load(fullSpecifier: string): Promise<void>

  // import induces the asynchronous load phase and then initializes
  // the given module and any of its transitive dependencies that
  // have not already begun initialization.
  async import(fullSpecifier: string): Promise<ModuleExportsNamespace>;

  // Necessary to thread a module exports namespace from this compartment into
  // the `moduleMap` Compartment constructor argument, without importing (and
  // consequently executing) the module.
  // TODO it may be possible to remove this method if module descriptors

>   // with the {specifier, compartment?} shape solve the same problem better.

Author: We (@erights, @patrick-soquet, @phoddie, @mhofman, myself) agreed to remove `module` from the proposal.

// The environment record does not contain a property for any names that are
// imported and reexported without a lexical binding.
// An `import name as alias` binding will have a property with the lexically bound alias.
// An `export name as alias` binding will have a property with the lexically bound name,
// whereas the module exports namespace will have a property with the alias.
interface ModuleEnvironmentRecord {}

// Bindings reflect the `import` and `export` statements of a module.
type Binding =
  { import: '*' | string, as?: string, from: string } |
  { export: '*' | string, as?: string, from?: string };

// Compartments support ECMAScript modules and linkage to other kinds of modules,
// notably allowing for JSON or WASM.
// These must provide an initializer function and may have bindings.
// The bindings correspond to the equivalent `import` and `export` declarations
// of an ECMAScript module.
type ThirdPartyStaticModuleRecord = {
  bindings?: Array<Binding>,
  // Initializes the module if it is imported.
  // Initialize may return a promise, indicating that the module uses
  // the equivalent of top-level-await.

>   initialize(environment: ModuleEnvironmentRecord, importMeta: Object | undefined),

Reviewer 2: ```suggestion
  initialize(environment: ModuleEnvironmentRecord, importMeta: Object | undefined, dynamicImport: (spec: string) => Promise<ModuleExportsNamespace>),
```

Author: @patrick-soquet I will follow-up with a change here that will interest you:

Jack brings up an interesting point that we need to thread `import` and `import.meta`. Consequently, we probably ought to have both `needsImport` and `needsImportMeta` flags on the third-party static-module-record. And since both are optional, the `initialize` function should probably take an options bag.

Reviewer 2: 😂 one more step forwards SystemJS format

Reviewer 2: 😂 one more step forwards SystemJS format

>   initialize(environment: ModuleEnvironmentRecord, importMeta: Object | undefined),

Reviewer 2: I currently understand and implement it as:

```ts
const env = {
    // writable: false, configurable: true
    // re-define when live-binding updates
    importLiveBindingNames: any,
    // getter, setter
    exportLiveBindingNames: any,
    __proto__: currentCompartment.[[globalThis]]
}
const env_proxy = new Proxy(env, onlyAllowGetSetTrap)
```

Please point out if there is RS error thanks!

If we cannot access `globalTHis` through `env` or any other API, then there is no way to compile a ES Module to a `StaticModuleRecord`.

  // Indicates that initialize needs to receive an importMeta.
  needsImportMeta: boolean,
};

// Static module records are an opaque token representing the compilation
// of a module that can be reused across multiple compartments.
interface StaticModuleRecord {
  // Static module records can be constructed from source.
  // XS allows third-party module records and source descriptors to
  // be precompiled as well.

>   constructor(source: string | { source: string } | ThirdPartyStaticModuleRecord);

Author: I’m definitely leaning toward taking this suggestion, again to minimize complications.

I am also wondering what we would lose if left out “pre-compiled third-party static-module-records” by reducing this constructor back to just taking the ESM string. cc @patrick-soquet

>   constructor(source: string | { source: string } | ThirdPartyStaticModuleRecord);

Reviewer 2: oh no, if we don't have "pre-compiled third-party static-module-records" then I have no way to use compartment, plz don't remove it

>   constructor(source: string | { source: string } | ThirdPartyStaticModuleRecord);

Reviewer 2: oh no, if we don't have "pre-compiled third-party static-module-records" then I have no way to use compartment, plz don't remove it

>   constructor(source: string | { source: string } | ThirdPartyStaticModuleRecord);

Author: I don’t mean to remove third-party static-module-records.

Currently, there are two ways to use a 3P-SMR in the proposed module descriptor:

```js
{ record: thirdParty }
{ record: new StaticModuleRecord(thirdParty) }
```

I’m suggesting that the latter form is redundant and I would like to remove it, unless there’s a benefit to the form in native implementations that I don’t know about yet.

>   constructor(source: string | { source: string } | ThirdPartyStaticModuleRecord);

Reviewer 2: Oh, I see. Then it's ok for me.

  // top-level-await for a module compiled from source.
  | {
    record: ThirdPartyStaticModuleRecord,

    // Properties to copy to the `import.meta` of the resulting module instance,
    // if the `record` has a `true` `needsImportMeta` property.
    importMeta?: Object,
  }

  // Use a static module record from the compartment in which this compartment
  // was constructed.
  // If the compartment _loads_ the corresponding module, the static
  // module record will be synchronously available only if it is already
  // in the parent compartment's synchronous memo.
  // Otherwise, loading induces the parent compartment to load the module,
  // but does not induce the parent compartment to load the module's transitive
  // dependencies.
  // This compartment then resolves the shallow dependencies according
  // to its own resolveHook and loads the consequent transitive dependencies.
  // If the compartment _imports_ the module, the behavior is equivalent
  // to loading for the retrieved static module record or third-party static
  // module record.
  | {
    record: string,

    // Properties to copy to the `import.meta` of the resulting module instance.
    importMeta?: Object,

>   }

Author: This form replaces a `string` type module descriptor, and allows us to distinguish `{ record: string }` from `{ instance: string }`.

If this idea is good, we could go farther and use `instance` instead of `namespace` so `{ source }` and `{ record }` consistently mean to use a `StaticModuleRecord` and every overload of `{ instance }` means to share a module instance (`ModuleEnvironmentRecord` and `ModuleExportsNamespace`).

  // FOR SHARED MODULE INSTANCES:

>   // To create an alias to an existing module instance in this compartment:

Reviewer 2: What do you mean by "this Compartment"? The compartment being constructed, or the incubator compartment of the constructing compartment?

  | {
    // A full specifier for another module in this compartment.
    instance: string,
  }

  // To create an alias to an existing module instance given the full specifier

>   // of the module in a different compartment:

Reviewer 2: What if the target compartment does not have the sync memo of this spec? Does it call the hook to load the module?

  | {
    instance: string,

    // A compartment instance.
    // We do not preclude the possibility that compartment is the same compartment,
    // as that is possible to achieve in moduleMapHook and loadHook, albeit not
    // in moduleMap.
    compartment: Compartment,
  }

  // To create an alias to an existing module instance given a module exports
  // namespace object that can pass a brand check:
  | {
    namespace: ModuleExportsNamespace,
  }

>   // If the given namespace does not pass a ModuleExportsNamespace brandcheck,

Author: No. We don’t propose to create a module exports namespace constructor, nor a module environment record constructor. Compartments generate entangled pairs of these objects internally.

  // the compartment will not support live bindings to the properties of that
  // object, but will instead produce an emulation of a module exports namespace,
  // using a frozen snapshot of the own properties of the given object.
  // The module exports namespace for this module does not reflect any future
  // changes to the shape of the given object.
  | {
    namespace: ^ModuleExportsNamespace,
  };

type CompartmentConstructorOptions = {
  // inherit, when true, indicates that this compartment shares the globals, global

>   // contour, and evaluators of the surrounding compartment.

Author: I think global contour might be an old name for global environment record https://tc39.es/ecma262/#sec-global-environment-records and I need to revisit my use of that term in this API sketch.

  // This would interact with the SES `lockdown` function: after lockdown,
  // a true would be incoherent and throw an exception.
  // When false, the compartment's globalThis would be populated only

>   // with a subset of the language's intrinsics and the compartment's

Author: Definitely not. I can’t imagine us permitting any object not currently in the SES-shim Lockdown permits by default. In order to avoid arguing about the exact list, my hope is that the only folks who care about `borrowGlobals: false` are those who in turn intend to also use Lockdown. `borrowGlobals: true` avoids that conversation in the cases where people are likely to object to the exact list of intrinsics in a `borrowGlobals: false` compartment.

  // or bindings array of a third-party StaticModuleRecord), to
  // the corresponding full specifier, given the full specifier of the
  // referring (containing) module.
  // The full specifier is the memo key for the module.
  // TODO This proposal does not yet account for import assertions,
  // which evidence shows are actually import type specifiers as they were
  // originally designed and may need to also participate in the memo key.
  resolveHook: (importSpecifier: string, referrerSpecifier: string) => string,

  // The compartment's load and import methods may require to load or initialize
  // additional modules.
  // If the compartment does not have the corresponding module on hand,
  // it will first consult the moduleMap for a descriptor of the needed module.
  moduleMap?: Record<string, ModuleDescriptor>,

  // If the moduleMap does not contain a descriptor for a necessary module,
  // it will then consult the moduleMapHook.
  // The moduleMapHook is necessary for matching computed specifiers,
  // like specifiers that match a prefix and have a corresponding
  // module descriptor.
  // The moduleMap and moduleMapHook are both synchronous.

>   moduleMapHook?: (fullSpecifier: string) => ModuleDescriptor?,

Author: It turns out that the Agoric-inspired `moduleMapHook` has the same signature and behavior as the Moddable-inspired `loadNowHook`. It might be an overall improvement in names if we:

- renamed `moduleMap` to simply `modules`,
- renamed `moduleMapHook` to `loadNowHook`, and
- kept `loadHook`
- kept `importNow`

`moduleMap` was not awkward when it was a positional argument, but now the so-called module-map is a module-descriptor-dictionary (neither module, nor Map), and now is the time to consider bike-shedding.

Author: We could also alternately use the precedent from `Array.fromAsync` and:

- rename `moduleMap` to `modules`,
- rename `moduleMapHook` and `loadNowHook` to simply `loadHook`
- rename `loadHook` to `loadAsyncHook`

Author: I’m going to remove `moduleMapHook` for now, with the intention of leaving open the possibility of a TC53 carve-out for `loadNowHook` later. For general JavaScript (with promises), and with module descriptors supported on both sync on moduleMap (`modules`) and returned by `loadHook`, I think we have everything we need.

  // to instantiate for every module, like Node.js's `import.meta.resolve`,
  // which can be avoided for modules that never utter `import.meta`
  // in their sources.
  // This hook gets called for any module that utters `import.meta`,
  // so some work can be deferred until just before the compartment
  // initializes the module.
  importMetaHook?: (fullSpecifier: string, importMeta: Object) => void,
};

interface Compartment {
  // Note: This single-argument form differs from earlier proposal versions,
  // implementations of SES shim, and Moddable's XS, which accept three arguments,
  // including a final options bag.
  constructor(options?: CompartmentConstructorOptions): Compartment;

  // Accessor for this compartment's globals.
  // If inherit is true, globalThis is object identical to the incubating
  // compartment's globalThis.
  // If inherit is false, globalThis is a unique, ordinary object
  // intrinsic to this compartment.
  // The globalThis object initially has only "shared intrinsics",
  // followed by compartment-specific "eval", "Function", and "Compartment",
  // followed by any properties transferred from the "globals"

>   // constructor option with the semantics of Object.assign.

Author: We could easily omit the `globals` option entirely since it’s duplicative with `compartment.globalThis`, which allows the holder to do whatever they want.

  // but cannot and should not emulate evaluation.
  async load(fullSpecifier: string): Promise<void>

  // import induces the asynchronous load phase and then initializes
  // the given module and any of its transitive dependencies that
  // have not already begun initialization.
  async import(fullSpecifier: string): Promise<ModuleExportsNamespace>;

  // TC53: some embedded systems hosts exclude promises and so require these
  // synchronous variants of import and load.
  // On hosts where these functions cannot succeed synchronously,
  // they must throw an error without the side effect of initializing any
  // additional modules.
  // For modules that have a top-level await, these would return
  // after the module first awaits but not after any subsequent promise queue
  // job.

  // If a host supports both importNow and import, they must share
  // a memo of full specifiers to promises for the module exports namespace.
  // importNow and loadNow followed by import or load must not accidentally
  // reinitialize the underlying module or produce different namespaces.

  importNow(fullSpecifier: string): ModuleExportsNamespace;

>   loadNow(fullSpecifier: string): void;

Reviewer 2: mark those two as normative optional?

This is the behavior of a dynamic import today, despite it being surprising.

We have chosen to embrace this hazard since it would be worse to have

> dynamic import and compartment import behave differently.

Author: I believe this is the most recent consensus, but defer to @bmeck to verify.

This is the behavior of a dynamic import today, despite it being surprising.

We have chosen to embrace this hazard since it would be worse to have
dynamic import and compartment import behave differently.

However, with `compartment.importNow`, a program can mitigate this hazard.
With `importNow`, the following program will not invoke the `then`
function exported by `'./thenable.js'`.

```js
await compartment.load('./thenable.js');
const thenableNamespace = compartment.importNow('./thenable.js');
```

[browserify]: https://browserify.org/
[import-map]: https://github.com/WICG/import-maps

> [jest-ses-interaction]: https://github.com/facebook/jest/issues/11952

Reviewer: ```suggestion
[jest]: https://www.npmjs.com/package/jest
[jest-ses-interaction]: https://github.com/facebook/jest/issues/11952
```
